messaging/src/main/java/io/atomix/messaging/impl/MessageDecoder.java

```diff
@@ -148,6 +185,29 @@ protected void decode(
     }
   }
 
+  static String readString(ByteBuf buffer, int length, Charset charset) {
+    if (buffer.isDirect()) {
+      final String result = buffer.toString(buffer.readerIndex(), length, charset);
+      buffer.skipBytes(length);
+      return result;
+    } else {
+      if (buffer.hasArray()) {
+        final int offset = buffer.arrayOffset() + buffer.readerIndex();
+        final String result;
+        if (offset == 0 && length == buffer.array().length) {
```

Comment:
IMHO, testing this is useless and you should just perform the second branch.

Reply:
removed.